### PR TITLE
Djm/autoplot

### LINKED ISSUE #2: Add plotting tools?
Should we add plotting tools to this package?  It's not clear to me:

- Whether it belongs in this package versus a separate one (that does purely visualization, say `epivis` or `epiplot`).
- Should be done the way we did it in `covidcast` (even if it gets put in its own package).  Maybe just facilities for choropleth plots, but with a stronger reliance on built-in mapping software (some of what I did in `covidcast` got a bit hairy...).

### PR DESCRIPTION
**Pull request by dajmcdon**
Adds `autoplot()` functionality and some helpers.

Needs #377 to pass checks.

* Closes #2 

See also
* cmu-delphi/forecasting-team-meta#15
* cmu-delphi/covid-hosp-forecast#94

### CODE REVIEW

File: R/epi_keys.R [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#' Grab any keys associated to an epi_df
+#'
+#' @param x a data.frame, tibble, or epi_df
+#' @param ... additional arguments passed on to methods
+#'
+#' @return If an `epi_df`, this returns all "keys". Otherwise `NULL`
+#' @keywords internal
+#' @export
+epi_keys <- function(x, ...) {
```
Comment (brookslogan) [inline R/epi_keys.R:9]: The term `epi_keys` seems to be [unintuitive](https://delphi-org.slack.com/archives/C028AUR6MV1/p1705521596770669) and [used in multiple ways in epipredict](https://delphi-org.slack.com/archives/C028AUR6MV1/p1696467136497469); and `ek` to refer to multiple things below.

Suggest:
- rename the `epi_keys` function `key_colnames` or maybe `key_names`

and maybe
- rename `ek` after killing time values to something like `geo_other_keynames`
- maybe rename `mv` and `ek` before killing time values to something more descriptive

Comment (brookslogan) [reply R/epi_keys.R:9]: If you'd like, I could try to do this renaming as well.

File: R/utils-arg.R [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            "Argument {.val {name}} must not be a missing value ({.val {NA}})."
+          )
+        }
+      }
+    }
+  )
+}
+
+arg_is_int <- function(..., allow_null = FALSE) {
+  arg_is_numeric(..., allow_null = allow_null)
```
Comment (nmdefries) [inline R/utils-arg.R:28]: suggestion: This function doesn't exist, and is causing an issue with trying to use `.max_facets`. I assume this was supposed to be `arg_is_scalar`.

```
autoplot(jhu_csse_daily_subset, case_rate_7d_av, .facet_by = "geo_value", .max_facets = 3)
# Error in arg_is_numeric(..., allow_null = allow_null) : 
#   could not find function "arg_is_numeric"
```

```suggestion
  arg_is_scalar(..., allow_null = allow_null)
```

File: R/autoplot.R [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+#'   were positions in the data frame, so expressions like `x:y` can
+#'   be used to select a range of variables.
+#' @param .color_by Which variables should determine the color(s) used to plot
+#'   lines. Options include:
+#'   * `all_keys` - the default uses the interaction of any key variables
+#'     including the `geo_value`
+#'   * `geo_value` - `geo_value` only
+#'   * `other_keys` - any available keys that are not `geo_value`
+#'   * `.response` - the numeric variables (same as the y-axis)
+#'   * `all` - uses the interaction of all keys and numeric variables
```
Comment (nmdefries) [inline R/autoplot.R:15]: issue: Getting an error using the "all" setting. I get the same error when trying to use `.facet_by = all`.

```
autoplot(aa, case_rate_7d_av, .color_by="all")
# Error in `dplyr::mutate()`:
# ℹ In argument: `.colours = switch(...)`.
# Caused by error in `interaction()`:
# ! object '.response_name' not found
```

Comment (nmdefries) [reply R/autoplot.R:15]: issue: `.response` and `none` generate the same plot, where the data is a single color. It sounds like `.response` is supposed to make higher values a different color than lower values.

File: R/autoplot.R [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+#'   * `none` - no coloring aesthetic is applied
+#' @param .facet_by Similar to `.color_by` except that the default is to display
+#'   each numeric variable on a separate facet
+#' @param .base_color Lines will be shown with this color. For example, with a
+#'   single numeric variable and faceting by `geo_value`, all locations would
+#'   share the same color line.
```
Comment (nmdefries) [inline R/autoplot.R:21]: question: Is this the correct usage?

```
autoplot(jhu_csse_daily_subset, case_rate_7d_av, .base_color = "red", .facet_by = "geo_value")
```

issue: The line color isn't changing when I do this, or if I try a hex color code.

File: R/autoplot.R [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    vars <- tidyselect::eval_select(names(allowed)[1], object)
+    cli::cli_warn(
+      "Plot variable was unspecified. Automatically selecting {.var {names(allowed)[1]}}."
+    )
+  } else { # if variables were specified, ensure that they are numeric
+    ok <- names(vars) %in% names(allowed)
+    if (!any(ok)) {
+      cli::cli_abort(
+        "None of the requested variables {.var {names(vars)}} are numeric."
+      )
```
Comment (nmdefries) [inline R/autoplot.R:68]: issue: I can't get this error to trigger. I have an epi_df with no numeric columns (sex is included as a key), but it gets plotted anyway.

test data

The `fake` field is plotted if I run either `autoplot(aa, fake)` or `autoplot(aa)`

Comment (nmdefries) [reply R/autoplot.R:68]: suggestion: Please add tests to check that warnings and errors in this function are raised appropriately.

**Review by nmdefries — CHANGES_REQUESTED** [review]
I'm unable to get a fair bit of `autoplot` behavior to work. Specifics below.

**Comment by nmdefries** [comment]
thought: Could be nice to be able to change the color scheme. That may be too feature-ful for `autoplot`, though.

File: DESCRIPTION [hunk 6]
```diff
@@ -72,6 +72,7 @@ Depends:
 URL: https://cmu-delphi.github.io/epiprocess/
 Collate: 
     'archive.R'
+    'autoplot.R'
     'correlation.R'
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply DESCRIPTION:76]: Yes on "out of scope". I checked with @brookslogan a while back, and there was a reason at the time that this was needed.

File: R/key_colnames.R [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  other_keys <- attr(x, "metadata")$other_keys
+  c("time_value", "geo_value", other_keys)
+}
+
+kill_time_value <- function(v) {
+  arg_is_chr(v)
+  v[v != "time_value"]
+}
+
+key_colnames_only <- function(x, ...) {
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/key_colnames.R:42]: Maybe better is `keys_no_time_value()`? We regularly need to grab the fields for `time_value`, `geo_value` and `$metadata$other_keys`. (This happens throughout `{epipredict}` as well). It's internal as far as I can tell. Any other ideas for names?

File: R/utils.R [hunk 8]
```diff
@@ -482,13 +482,13 @@ quiet <- function(x) {
...
-  n <- as.character(sys.call())[-1]
-  if (!is.null(n0 <- names(x))) {
-    n[n0 != ""] <- n0[n0 != ""]
-  }
-  names(x) <- n
-  return(x)
+  # converted to thin wrapper around
+  rlang::dots_list(
+    ...,
+    .homonyms = "error",
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/utils.R:488]: Yes. Safety preferred here. Typical usage is to construct the list for a function return. So duplicate names should error.

File: R/utils-arg.R [hunk 9]
```diff
@@ -0,0 +1,62 @@
+handle_arg_list <- function(..., tests) {
+  values <- list(...)
+  names <- eval(substitute(alist(...)))
+  names <- purrr::map(names, deparse)
+
+  purrr::walk2(names, values, tests)
+}
+
+arg_is_scalar <- function(..., allow_null = FALSE, allow_na = FALSE) {
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/utils-arg.R:9]: Added as #403. Should apply to `{epipredict}` as well.

File: R/autoplot.R [hunk 10]
```diff
@@ -0,0 +1,156 @@
+#' Automatically plot an epi_df
+#'
+#' @param object An `epi_df`
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/autoplot.R:3]: Possibly, though worth brainstorming what we'd want it to look like.

File: R/utils-arg.R (hunk 2 above)
Comment (dajmcdon) [reply R/utils-arg.R:28]: No, this is correct. Adding the missing function.

File: R/autoplot.R (hunk 4 above)
Comment (dajmcdon) [reply R/autoplot.R:21]: No. The default colouring is controlled by `.color_by`. So you'd need:
```r
autoplot(jhu_csse_daily_subset, case_rate_7d_av, .color_by = "none", .base_color = "red", .facet_by = "geo_value")
```

File: R/key_colnames.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/key_colnames.R:42]: `keys_no_time_value` sounds fine. `key_colnames_no_time_value` better parallels `key_colnames` but is a bit long.

File: R/autoplot.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/autoplot.R:3]: Opened https://github.com/cmu-delphi/epiprocess/issues/404

File: R/autoplot.R (hunk 5 above)
Comment (dajmcdon) [reply R/autoplot.R:75]: I fixed this bug, but I don't think I have the bandwidth to test all the branches. Perhaps someone else can give it a go?

Comment (nmdefries) [reply R/autoplot.R:75]: Sure, I can do that.

File: R/autoplot.R (hunk 3 above)
Comment (dajmcdon) [reply R/autoplot.R:15]: I added it as an example.

File: R/autoplot.R (hunk 4 above)
Comment (dajmcdon) [reply R/autoplot.R:21]: Added this as an example.

File: DESCRIPTION (hunk 6 above)
[the remark this answers is not in the document]
Comment (brookslogan) [reply DESCRIPTION:76]: (We need Collate to get roxygen to order parts of multi-function topics in the way we want when the functions are in multiple files.)

**Comment by dajmcdon** [comment]
@brookslogan Trying to document to regenerate the NAMESPACE. I'm hitting:

```r
devtools::document()
ℹ Updating epiprocess documentation
ℹ Loading epiprocess
Error: 'archive_cases_dv_subset' is not an exported object from 'namespace:epiprocess'
```

Followed by:
```r
devtools::document()
ℹ Updating epiprocess documentation
ℹ Loading epiprocess
Error in `withCallingHandlers()` at epiprocess/R/data.R:137:7:
! An error was raised while attempting to evaluate a promise (prepared
with `delayed_assign_with_unregister_awareness`) while an `unregister` or
`unregister_namespace` call was being evaluated. This can happen, for example,
when `devtools::load_all` reloads a package that contains a buggy promise,
because reloading can cause old package-level promises to be forced via
`pkgload::unregister` and `pkgload:::unregister_namespace`, due to
https://github.com/r-lib/pkgload/pull/157. If this is the current situation,
you might be able to be successfully reload the package again after
`unloadNamespace`-ing it (but this situation will keep re-occurring every
other `devtools::load` and every `devtools:document` until the bug or
situation generating the promise's error has been resolved).
Caused by error in `as.data.table()`:
! could not find function "as.data.table"
Run `rlang::last_trace()` to see where the error occurred.
```

**Review by nmdefries — APPROVED** [review]

**Pull request merged**